include/lsst/afw/image/ExposureInfo.h

```diff
@@ -213,6 +213,125 @@ class ExposureInfo final {
 
+    /**
+     * Add a generic component to the ExposureInfo.
+     *
+     * If another component is already present under the key, it is
+     * overwritten. If a component of a different type is present under the
+     * same name, this method raises an exception.
+     *
+     * @tparam T a subclass of typehandling::Storable
+     *
+     * @param key a strongly typed identifier for the component
+     * @param object the object to add.
+     *
+     * @throws pex::exceptions::TypeError Thrown if a component of a
+     *      different type is present under the requested name.
+     * @throws pex::exceptions::RuntimeError Thrown if the insertion failed for
+     *      implementation-dependent reasons.
+     * @exceptsafe Provides basic exception safety (a pre-existing component
+     *             may be removed).
+     *
+     * @note if `object` is a null pointer, then `hasComponent(key)` shall
+     *       return `false` after this method returns. This is for
+     *       compatibility with old ExposureInfo idioms, which often use
+     *       assignment of null to indicate no data.
+     */
+    // non-shared_ptr components are incompatible with table::io,
+    // they may be supported later
+    template <class T>
+    void setComponent(typehandling::Key<std::string, std::shared_ptr<T>> const& key,
+                      std::shared_ptr<T> const& object) {
+        static_assert(std::is_base_of<typehandling::Storable, T>::value, "T must be a Storable");
+        _setStorableComponent(key, object);
+    }
+
+    /**
+     * Add a generic component to the ExposureInfo.
+     *
+     * If another component is already present under the key, it is
+     * overwritten. If a component of a different type is present under the
+     * same name, this method raises an exception.
+     *
+     * @param key an identifier for the component
+     * @param object the object to add.
+     * @returns a strongly-typed key that can be used to retrieve the component
+     *
+     * @throws pex::exceptions::TypeError Thrown if a component of a
+     *      different type is present under the requested name.
+     * @throws pex::exceptions::RuntimeError Thrown if the insertion failed for
+     *      implementation-dependent reasons (e.g., `object` could not
+     *      be copied).
+     * @exceptsafe Provides basic exception safety (a pre-existing component
+     *             may be removed).
+     *
+     * @note if `T` is a pointer type and `object` is null, then
+     *       `hasComponent([return value])` shall return `false` after this
+     *       method returns. This is for compatibility with old ExposureInfo
+     *       idioms, which often use assignment of null to indicate no data.
+     */
+    template <class T>
+    typehandling::Key<std::string, T> setComponent(std::string const& key, T const& object) {
+        auto typedKey = typehandling::makeKey<T>(key);
+        setComponent(typedKey, object);
+        return typedKey;
+    }
```

Comment:
What do you think about removing this overload, in order to encourage a convention (which we'd also have to document somewhere) that packages that add new components to `ExposureInfo` should defined a module/namespace-level `Key` constant that should be used to access it?  That would cut down on the use of string literals, add a bit of type safety, and simplify the `ExposureInfo` interface a bit.

Reply:
I'm surprised to hear you request this, since having such an adder for `MutableGenericMap` was your idea. But I'm fine with removing it (it's redundant in Python anyway).

Reply:
Yeah, I suppose it's just the slight distinction between the roles of a bare `GenericMap` and `ExposureInfo` that swings me in opposite directions.  It isn't a terribly strongly preference in either case.